pandas/core/indexes/interval.py

```diff
@@ -609,6 +609,8 @@ def _maybe_convert_i8(self, key):
         if scalar:
             # Timestamp/Timedelta
             key_dtype, key_i8 = infer_dtype_from_scalar(key, pandas_dtype=True)
+            if lib.is_period(key):
+                key_i8 = key.ordinal
```

Comment:
hmm what if you do this in the return key_i8 (e.g. inside `infer_dtype_from_scalar`), I *think* we do with for Timestamp/Timedelta?

Reply:
Yeah, that's how it was before. I didn't realize that we return key_i8 for Timestamp/Timedelta, so thanks for pointing that out.

IMO, it makes a lot more sense for the `val` returned from `infer_dtype_from_scalar` to be have the dtype of the returned `dtype` (ie returning `Period(2020-01), 'period[M]'` as opposed to `600, 'period[M]'`, since 600 is an int, not 'period[M]'). But I'm willing to do it either way

Reply:
Despite my opinion above, I think we have two options (using example of `infer_dtype_from_scalar(Period('2020-01'))`):

1. Returning `val=Period('2020-01')`, in which case we need to convert `key` to `key.ordinal` in this Interval code
2. Returning `val=600`, in which case we'd need to convert the ordinal value back to a Period in the `sanitize_array` function

What would you prefer @jreback? If you prefer 2, how would I convert from an ordinal value back to a Period?

Reply:
@jreback What do you think here?

Reply:
yes I think restore the code in pandas/core/dtypes/cast.py and remove the added code here and it should just work

Reply:
Doing that would break the [sanitize_array function](https://github.com/pandas-dev/pandas/pull/35966/files#diff-65590f3c100517411c1217385d9c68f6R471) as it is expecting the returned value of `infer_dtype_from_scalar` to be the Period, not the Period.ordinal value.

But if, in the `sanitize_array` function, I convert the ordinal value back into a Period then everything should work. This would be essentially the opposite of the logic above:
```python
if lib.is_period(key):
    key = ordinal_to_period(ordinal)
```

@jreback I'm fine doing this, but idk a way to do that `ordinal_to_period`, do you?

Reply:
this won't break sanitize_array as it feeds back to construct_1d_arraylike_from_scalar which can take a Period (or *should* be able to).

we don't want to convert back and forth.

Reply:
@jreback do you want me to make this change? If so, how do i convert from ordinal to period?

Reply:
no we do not want to covert back
rather it shouldn't convert to ordinal in the first place (which is what the original code was doing)

Reply:
Correct, the original code in `sanitize_array` called
```python
dtype, value = infer_dtype_from_scalar(value)
```
Which does not convert a Period to it's ordinal value, but it also would return `dtype="object"` which is the why Period columns would be typed as "object" instead of "Period". To correct this error, I changed that code to
```python
dtype, value = infer_dtype_from_scalar(value, pandas_dtype=True)
```
@jreback Are you suggesting that `infer_dtype_from_scalar` shouldn't convert to ordinal in the first place or that I shouldn't add `pandas_dtype=True` in the first place?

Reply:
> Are you suggesting that infer_dtype_from_scalar(..., pandas_dtype=True) shouldn't convert to ordinal in the first place

yes

Reply:
Haha, @jreback I wasn't expecting you to pick that one and made changes in the opposite direction (I will revert though).

I totally agree that `infer_dtype_from_scalar(..., pandas_dtype=True)` shouldn't convert to ordinal in the first place, but if that's the case, then
```python
if lib.is_period(key):
    key_i8 = key.ordinal
```
in `pandas/core/indexes/interval.py` is still necessary

Reply:
Reverted, are we good then? Or are you still looking for something else in this PR?